Vasyalvv short links

Vasyalvv opened the pull request:
Доработка замечаний fix #71.

- Исправлены недочеты в коде;
- Исправлены проблемы возникшие при предыдущем слиянии;
- Метод создания короткой ссылки возвращает Id созданной ссылки;
- Создан отдельный метод получения короткой ссылки по Id;
- Методы Web API создания короткой ссылки и получения короткой ссылки по Id возвращают результат CreatedAtAction и AcceptedAtAction соответственно;
- Маршруты контроллера Web-приложения приведены в порядок, в соответствии с заданием. Теперь доступ к действиям контроллера происходит по стандартному маршруту "https://localhost/url/action/id?", переход по короткой ссылке "https://localhost/url/17D9...", создание короткой ссылки по POST запросу "https://localhost/url/" с передачей исходной ссылки в теле запроса.

Vasyalvv committed:
Испарвление ошибок CPP-подхода

Vasyalvv committed:
Исправлена проблема с кодировкой

Vasyalvv committed:
Устранение проблем слияния

Vasyalvv committed:
Прменение версионирования для Web API сервиса коротких ссылок

Vasyalvv committed:
Метод создания короткой ссылки возвращает идентификатор созданной ссылки.
Отдельный метод для получения хеша короткой ссылки по идентификатору.

Vasyalvv committed:
Изменены типы возвращаемых значений методов Web API

Vasyalvv committed:
Изменен путь к WebAPI, в путь добавлена версия API

Vasyalvv committed:
Приведены в порядок маршруты контроллера. Теперь доступ к действиям контроллера происходит по стандартному маршруту "https://localhost/url/action/id?", переход по короткой ссылке "https://localhost/url/17D9...", создание короткой ссылки по POST запросу "https://localhost/url/" с передачей ссылки в теле запроса

Infarh commented:
1. Сократить сокрощённую ссылку до 5 символов хеша (количество символов должно указываться в файле конфигурации)
2. Сокращённая ссылка должна одинаково работать для любой длины строки. Поиск в БД должен осуществляться по совпадению начала строки хеша

Vasyalvv committed:
В Web API контроллере методы в случае неудачи возвращают NotFound. Метод AddUrl при неудаче возвращает BadRequest

Vasyalvv committed:
Ограничение количества символов для короткой ссылки. Для короткой ссылки в БД сохраняется MD5-хеш из 32 символов - это является псевдонимом ссылки. Для получения оригинальной ссылки в запросе к WebAPI можно передавать как псевдоним целиком (api/v1/url?Alias=D0D16672757FADB7FA0056C62114C533), так и начальную его часть (api/v1/url?Alias=D0D16), поиск в БД происходит по совпадению начала строки.
При генерировании псевдонима, по умолчанию возвращается полный псевдоним 32 символа. Если такая длина избыточна, то ее можно ограничить указав параметр "ShortUrlMaxLength" в файле appsettings.json. Либо это ограничение можно указать в запросе к WebAPI (api/v1/url/alias/1084?Length=7).

Vasyalvv committed:
Добавлена статистика по использованию коротких ссылок

Infarh reviewed (changes requested):
Добавить проверку и коррекцию ввода пользователя ссылки - добавление в начало схемы адреса URL в контроллерах
```C#
if (!Regex.IsMatch(url, @"^https?://"))
    url = "http://" + url;
```

В сервисах добавить валидацию адреса и генерацию исключения
```C#
if (!Regex.IsMatch(Url, @"^https?://"))
    throw new FormatException("Строка адреса не имеет схемы");
```

Infarh commented:
Также поправить текст "обновить" на "сбросить" в окне статистики

Vasyalvv committed:
Исправлены подписи.
Добавлена проверка и дополнение Url-адреса
